### PR TITLE
Make libnetdata headers compilable by C++.

### PR DESCRIPTION
**Pull request by KickerTom**
Minimal set of changes to libnetdata to allow including it from C++ code. Please note that I haven't even tried compiling netdata itself with these changes, nor I checked the code can be really used from C++ code. This only removes compilation errors and serves for demonstration purposes.
Please refer to issue #10178 for more details.

### CODE REVIEW

**Comment by KickerTom** [comment]
Seems to be complete now. Fixes #10178.
Please note I have added the 'extern "C"' only in the main libnetdata.h and log/log.h headers. If some other headers are expected to be used alone, without including the whole lot, let me know, I will add it there too.

**Review by thiagoftsm — DISMISSED** [review]
I tested your PR now and I could not identify any problem neither with the plugins nor with stream, that are direct affected from these changes.

Thank you @KickerTom  for your contribution!

File: libnetdata/dictionary/dictionary.h [hunk 1]
```diff
@@ -13,7 +13,7 @@ struct dictionary_stats {
 };
 
 typedef struct name_value {
-    avl avl;                // the index - this has to be first!
+    avl avl_instance;       // the index - this has to be first!
```
Comment (mfundul) [inline libnetdata/dictionary/dictionary.h:16]:
```suggestion
    avl avl_node;           // the index - this has to be first!
```
Let's keep it consistent with the similar change above.

File: libnetdata/avl/avl.h [hunk 2]
```diff
@@ -34,13 +34,13 @@ typedef struct avl {
...
-typedef struct avl_tree {
+typedef struct avl_tree_type {
     avl *root;
     int (*compar)(void *a, void *b);
-} avl_tree;
+} avl_tree_type;
 
 typedef struct avl_tree_lock {
-    avl_tree avl_tree;
+    avl_tree_type avl_tree;
```
Comment (mfundul) [inline libnetdata/avl/avl.h:43]:
```suggestion
    avl_tree avl_tree_instance;
```
I don't like how changes propagated outside of libnetdata. Let's go the other way around, similar to what you did with `struct avl`.

**Review by mfundul — CHANGES_REQUESTED** [review]

File: libnetdata/avl/avl.h (hunk 2 above)
Comment (KickerTom) [reply libnetdata/avl/avl.h:43]: I originally tried that as well. But the avl_tree member is being accessed on many places, so it propagates outside even more. Moreover, as part of the app config structure (appconfig.h, struct config) it is being used and accessed often by collector plugins, so changing the member name has bigger potential to break third-party plugins than the type name change.
But both variants are possible, so if you think changing member name is better, I can do that.

Comment (mfundul) [reply libnetdata/avl/avl.h:43]: No, you are right, there are a bunch of initializers that touch the `avl_tree` field at various places. It's probably not worth it.

**Review by mfundul — APPROVED** [review]

**Review by thiagoftsm — APPROVED** [review]

**Comment by thiagoftsm** [comment]
Thank you very much @KickerTom !

**Pull request merged**